iocore/eventsystem/UnixEventProcessor.cc

```diff
@@ -32,399 +32,229 @@
 #include "ts/ink_defs.h"
 #include "ts/hugepages.h"
 
+#include <algorithm>
+
 /// Global singleton.
 class EventProcessor eventProcessor;
 
-class ThreadAffinityInitializer : public Continuation
-{
-  typedef ThreadAffinityInitializer self;
-
-public:
-  /// Default construct.
-  ThreadAffinityInitializer() { SET_HANDLER(&self::set_affinity); }
-  /// Load up basic affinity data.
-  void init();
-  /// Set the affinity for the current thread.
-  int set_affinity(int, Event *);
-  /// Allocate a stack.
-  /// @internal This is the external entry point and is different depending on
-  /// whether HWLOC is enabled.
-  void *alloc_stack(EThread *t, size_t stacksize);
-
-protected:
-  /// Allocate a hugepage stack.
-  /// If huge pages are not enable, allocate a basic stack.
-  void *alloc_hugepage_stack(size_t stacksize);
-
 #if TS_USE_HWLOC
-
-  /// Allocate a stack based on NUMA information, if possible.
-  void *alloc_numa_stack(EThread *t, size_t stacksize);
-
-private:
-  hwloc_obj_type_t obj_type;
-  int obj_count        = 0;
-  char const *obj_name = nullptr;
+static const hwloc_obj_type_t kAffinity_objs[] = 
+   { HWLOC_OBJ_MACHINE, HWLOC_OBJ_NUMANODE, HWLOC_OBJ_SOCKET, HWLOC_OBJ_CORE
+#if HAVE_HWLOC_OBJ_PU
+         , HWLOC_OBJ_PU   // include only if valid in HWLOC version
 #endif
-};
+  };
 
-ThreadAffinityInitializer Thread_Affinity_Initializer;
+static const char *const kAffinity_obj_names[] =
+   { "[Unrestricted]",         "NUMA Node",        "Socket",         "Core"
+         , "Logical CPU"  // [ignore if not found]
+   };
 
-namespace
+// Pretty print our CPU set
+void pretty_print_cpuset(const char *thrname, hwloc_obj_type_t objtype, int affid, hwloc_const_cpuset_t cpuset)
 {
-/// This is a wrapper used to convert a static function into a continuation. The function pointer is
-/// passed in the cookie. For this reason the class is used as a singleton.
-/// @internal This is the implementation for @c schedule_spawn... overloads.
-class ThreadInitByFunc : public Continuation
-{
-public:
-  ThreadInitByFunc() { SET_HANDLER(&ThreadInitByFunc::invoke); }
-  int
-  invoke(int, Event *ev)
-  {
-    void (*f)(EThread *) = reinterpret_cast<void (*)(EThread *)>(ev->cookie);
-    f(ev->ethread);
-    return 0;
-  }
-} Thread_Init_Func;
-}
+  auto n = std::find( std::begin(kAffinity_objs), std::end(kAffinity_objs), objtype ) - std::begin(kAffinity_objs);
+  n %= countof(kAffinity_objs); // remap end to index zero
```

Comment:
There may be easier ways to accomplish wrapping the end to 0 than re-writing `countof`.